File: analyze_categories.js
#!/usr/bin/env node
/**
 * Kategori Analiz Scripti
 * Tüm yorumları kategorilere göre gruplar, sayar ve sıralar
 */

const fs = require('fs');
const path = require('path');

// Kategorileri yükle
const categoriesData = JSON.parse(
  fs.readFileSync('data/categories.json', 'utf-8')
);
const CATEGORIES = {};
categoriesData.categories.forEach(cat => {
  CATEGORIES[cat.id] = cat;
});

// Kategori sayaçları
const categoryCounts = {};

// Tüm JSON dosyalarını bul
const dataDir = 'data';
const jsonFiles = fs.readdirSync(dataDir)
  .filter(file => file.endsWith('-full.json'))
  .sort();

console.log('\n' + '='.repeat(80));
console.log('KATEGORİ ANALİZ RAPORU');
console.log('='.repeat(80) + '\n');
console.log(`Taranan dosya sayısı: ${jsonFiles.length}\n`);

let totalComments = 0;

// Her dosyayı işle
jsonFiles.forEach(filename => {
  const filepath = path.join(dataDir, filename);
  const data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
  
  // Issues içindeki yorumları say
  if (data.issues) {
    data.issues.forEach(issue => {
      if (issue.comments) {
        issue.comments.forEach(comment => {
          const category = comment.category || 'other';
          categoryCounts[category] = (categoryCounts[category] || 0) + 1;
          totalComments++;
        });
      }
    });
  }
});

// Kategorileri sayıya göre sırala (en çoktan en aza)
const sortedCategories = Object.entries(categoryCounts)
  .sort((a, b) => b[1] - a[1]);

// Sonuçları yazdır
console.log(`Toplam yorum sayısı: ${totalComments}\n`);
console.log('='.repeat(80));
console.log(
  'SIRA'.padEnd(6) + 
  'KATEGORİ ID'.padEnd(25) + 
  'KATEGORİ ADI'.padEnd(30) + 
  'YORUM SAYISI'.padEnd(15) + 
  'YÜZDE'
);
console.log('='.repeat(80));

const results = [];
sortedCategories.forEach(([catId, count], index) => {
  const rank = index + 1;
  const catName = CATEGORIES[catId]?.name || catId;
  const percentage = totalComments > 0 ? (count / totalComments * 100) : 0;
  
  console.log(
    String(rank).padEnd(6) +
    catId.padEnd(25) +
    catName.padEnd(30) +
    String(count).padEnd(15) +
    percentage.toFixed(2).padStart(6) + '%'
  );
  
  results.push({
    rank: rank,
    category_id: catId,
    category_name: catName,
    count: count,
    percentage: Math.round(percentage * 100) / 100
  });
});

console.log('='.repeat(80) + '\n');

// Sonuçları JSON'a kaydet
const outputData = {
  generated_at: new Date().toISOString(),
  total_comments: totalComments,
  total_files_analyzed: jsonFiles.length,
  categories: results
};

const reportsDir = 'reports';
if (!fs.existsSync(reportsDir)) {
  fs.mkdirSync(reportsDir, { recursive: true });
}

const outputFile = path.join(reportsDir, 'category_analysis.json');
fs.writeFileSync(outputFile, JSON.stringify(outputData, null, 2), 'utf-8');

console.log(`✅ Analiz sonuçları kaydedildi: ${outputFile}\n`);

// Export for use in other scripts
module.exports = { results, totalComments };

File: analyze_categories.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Kategori Analiz Scripti
Tüm yorumları kategorilere göre gruplar, sayar ve sıralar
"""

import json
from pathlib import Path
from collections import defaultdict
from datetime import datetime

def analyze_categories():
    """Tüm dosyalardaki yorumları kategorilere göre analiz et"""
    
    # Kategorileri yükle
    with open('data/categories.json', 'r', encoding='utf-8') as f:
        categories_data = json.load(f)
        CATEGORIES = {cat['id']: cat for cat in categories_data['categories']}
    
    # Kategori sayaçları
    category_counts = defaultdict(int)
    total_comments = 0
    
    # Tüm JSON dosyalarını tara
    data_dir = Path('data')
    json_files = sorted(data_dir.glob('*-full.json'))
    
    print(f"\n{'='*80}")
    print("KATEGORİ ANALİZ RAPORU")
    print(f"{'='*80}\n")
    print(f"Taranan dosya sayısı: {len(json_files)}\n")
    
    # Her dosyayı işle
    for json_file in json_files:
        with open(json_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Issues içindeki yorumları say
        if 'issues' in data:
            for issue in data['issues']:
                if 'comments' in issue:
                    for comment in issue['comments']:
                        category = comment.get('category', 'other')
                        category_counts[category] += 1
                        total_comments += 1
    
    # Kategorileri sayıya göre sırala (en çoktan en aza)
    sorted_categories = sorted(category_counts.items(), key=lambda x: x[1], reverse=True)
    
    # Sonuçları yazdır
    print(f"Toplam yorum sayısı: {total_comments}\n")
    print(f"{'='*80}")
    print(f"{'SIRA':<6} {'KATEGORİ ID':<25} {'KATEGORİ ADI':<30} {'YORUM SAYISI':<15} {'YÜZDE':<10}")
    print(f"{'='*80}")
    
    results = []
    for rank, (cat_id, count) in enumerate(sorted_categories, 1):
        cat_name = CATEGORIES.get(cat_id, {}).get('name', cat_id)
        percentage = (count / total_comments * 100) if total_comments > 0 else 0
        
        print(f"{rank:<6} {cat_id:<25} {cat_name:<30} {count:<15} {percentage:>6.2f}%")
        
        results.append({
            'rank': rank,
            'category_id': cat_id,
            'category_name': cat_name,
            'count': count,
            'percentage': round(percentage, 2)
        })
    
    print(f"{'='*80}\n")
    
    # Sonuçları JSON'a kaydet
    output_data = {
        'generated_at': datetime.now().isoformat(),
        'total_comments': total_comments,
        'total_files_analyzed': len(json_files),
        'categories': results
    }
    
    output_file = Path('reports/category_analysis.json')
    output_file.parent.mkdir(exist_ok=True)
    
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(output_data, f, ensure_ascii=False, indent=2)
    
    print(f"✅ Analiz sonuçları kaydedildi: {output_file}\n")
    
    return results, total_comments

if __name__ == "__main__":
    analyze_categories()
